### PR TITLE
New class SIMmodel

### PR DESCRIPTION
**Pull request by kmokstad**
A new class SIMmodel is introduced containing those parts of SIMbase that dealt with model input and generation, including initial condition handling. This will require some minor modifications of those apps using initial condtions + some regression test updates due to change in console output.

Have so far only tested it for Elasticity and FractureElasticity, but please have a look.

### CODE REVIEW

File: Apps/Common/SIMNodalConstraint.h [hunk 1]
```diff
@@ -77,18 +70,12 @@ void NodalConstraintProcessor<SIM3D>::apply(SIM3D& sim, TopologySet& myEntitys,
...
-  //! \brief Default constructor.
-  //! \param[in] n1 Dimension of the primary solution field
-  //! \param[in] check If \e true, ensure the model is in a right-hand system
-  SIMNodalConstraint(const SIMbase::CharVec& unf,
-                     bool checkRHS=false) :
-    Dim(unf,checkRHS) {}
-
-  //! \brief Empty destructor
+  //! \brief The constructor forwards to the parent class constructor.
+  SIMNodalConstraint(const std::vector<unsigned char>& unf) : Dim(unf) {}
```
Comment (akva2) [inline Apps/Common/SIMNodalConstraint.h]: this means you can never use checkRHS on models with nodal constraints.

File: src/SIM/ModelGenerator.C [hunk 2]
```diff
@@ -62,32 +64,29 @@ std::string DefaultGeometry1D::createG2 (int nsd) const
...
-    IFEM::cout <<"\tX1 = "<< XYZ << std::endl;
+    IFEM::cout <<"\n\tX1 = "<< XYZ;
     g2.append(XYZ);
   }
   else
   {
     XYZ = "1.0";
     if (utl::getAttribute(geo,"L",XYZ))
-      IFEM::cout <<"  Length scale: "<< XYZ << std::endl;
+      IFEM::cout <<"\n\tLength: "<< XYZ;
```
Comment (akva2) [inline src/SIM/ModelGenerator.C]: can we make this consistent (i.e. use = here as well) please ?

File: src/SIM/ModelGenerator.C [hunk 3]
```diff
@@ -106,61 +105,66 @@ TopologySet DefaultGeometry1D::createTopologySets (const SIMbase&) const
...
   g2.append("\n2 2\n0 0 1 1\n2 2\n0 0 1 1");
 
   Vec3 X0;
   std::string corner;
-  if (utl::getAttribute(geo,"X0",corner)) {
+  if (utl::getAttribute(geo,"X0",corner))
+  {
     std::stringstream str(corner); str >> X0;
-    IFEM::cout <<"  Corner: "<< X0 << std::endl;
+    IFEM::cout <<"\n\tCorner: "<< X0;
```
Comment (akva2) [inline src/SIM/ModelGenerator.C]: = please

File: src/SIM/ModelGenerator.C [hunk 4]
```diff
@@ -106,61 +105,66 @@ TopologySet DefaultGeometry1D::createTopologySets (const SIMbase&) const
...
   }
 
   double scale = 1.0;
   if (utl::getAttribute(geo,"scale",scale))
-    IFEM::cout <<"  Scale: "<< scale << std::endl;
+    IFEM::cout <<"\n\tScale: "<< scale;
```
Comment (akva2) [inline src/SIM/ModelGenerator.C]: =

File: src/SIM/ModelGenerator.C [hunk 5]
```diff
@@ -106,61 +105,66 @@ TopologySet DefaultGeometry1D::createTopologySets (const SIMbase&) const
...
 
   double Lx = 1.0, Ly = 1.0;
   if (utl::getAttribute(geo,"Lx",Lx))
-    IFEM::cout <<"  Length in X: "<< Lx << std::endl;
+    IFEM::cout <<"\n\tLength in X: "<< Lx;
```
Comment (akva2) [inline src/SIM/ModelGenerator.C]: =

File: src/SIM/ModelGenerator.C [hunk 6]
```diff
@@ -106,61 +105,66 @@ TopologySet DefaultGeometry1D::createTopologySets (const SIMbase&) const
...
   Lx *= scale;
   if (utl::getAttribute(geo,"Ly",Ly))
-    IFEM::cout <<"  Length in Y: "<< Ly << std::endl;
+    IFEM::cout <<"\n\tLength in Y: "<< Ly;
```
Comment (akva2) [inline src/SIM/ModelGenerator.C]: =

File: src/SIM/ModelGenerator.C [hunk 7]
```diff
@@ -206,17 +211,17 @@ std::string DefaultGeometry3D::createG2 (int) const
 
   double scale = 1.0;
   if (utl::getAttribute(geo,"scale",scale))
-    IFEM::cout <<"\tscale = "<< scale << std::endl;
+    IFEM::cout <<"\n\tScale = "<< scale;
 
   double Lx = 1.0, Ly = 1.0, Lz = 1.0;
   if (utl::getAttribute(geo,"Lx",Lx))
-    IFEM::cout <<"\tLength in X: "<< Lx << std::endl;
+    IFEM::cout <<"\n\tLength in X: "<< Lx;
```
Comment (akva2) [inline src/SIM/ModelGenerator.C]: =

File: src/SIM/ModelGenerator.C [hunk 8]
```diff
@@ -206,17 +211,17 @@ std::string DefaultGeometry3D::createG2 (int) const
...
   Lx *= scale;
   if (utl::getAttribute(geo,"Ly",Ly))
-    IFEM::cout <<"\tLength in Y: "<< Ly << std::endl;
+    IFEM::cout <<"\n\tLength in Y: "<< Ly;
```
Comment (akva2) [inline src/SIM/ModelGenerator.C]: =

File: src/SIM/ModelGenerator.C [hunk 9]
```diff
@@ -206,17 +211,17 @@ std::string DefaultGeometry3D::createG2 (int) const
...
   Ly *= scale;
   if (utl::getAttribute(geo,"Lz",Lz))
-    IFEM::cout <<"\tLength in Z: "<< Lz << std::endl;
+    IFEM::cout <<"\n\tLength in Z: "<< Lz;
```
Comment (akva2) [inline src/SIM/ModelGenerator.C]: =

File: src/SIM/ModelGenerator.C [hunk 10]
```diff
@@ -227,11 +232,12 @@ std::string DefaultGeometry3D::createG2 (int) const
...
 
   std::string corner;
-  if (utl::getAttribute(geo,"X0",corner)) {
+  if (utl::getAttribute(geo,"X0",corner))
+  {
     std::stringstream str(corner);
     Vec3 X0;
     str >> X0;
-    IFEM::cout <<"\tCorner: "<< X0 << std::endl;
+    IFEM::cout <<"\n\tCorner: "<< X0;
```
Comment (akva2) [inline src/SIM/ModelGenerator.C]: =

File: src/SIM/Test/TestInitialConditions.C [hunk 11]
```diff
@@ -14,20 +14,35 @@
...
     ASSERT_FLOAT_EQ((float)i,(*sim.getSclFunc(i))(Vec3()));
+
   // Initial conditions
-  ASSERT_TRUE(sim.getICs().begin() != sim.getICs().end());
-  const std::vector<SIMdependency::ICInfo>& bar = sim.getICs().begin()->second;
-  for (std::vector<SIMdependency::ICInfo>::const_iterator info = bar.begin();
-       info != bar.end(); info++)
+  const SIMmodel::InitialCondVec* ic = sim.getIC();
+  ASSERT_TRUE(ic != nullptr);
+  for (auto info = ic->begin(); info != ic->end(); ++info)
```
Comment (akva2) [inline src/SIM/Test/TestInitialConditions.C]: for (auto info : *ic) kanskje ?

File: src/Utility/Functions.C [hunk 12]
```diff
@@ -434,17 +438,15 @@ ScalarFunc* utl::parseTimeFunc (const char* func, const std::string& type)
...
 {
   if (func.empty()) return nullptr;
 
-  if (print)
-    IFEM::cout <<": ";
+  IFEM::cout <<": ";
   Real p = Real(0);
   if (type == "expression")
   {
-    if (print)
```
Comment (akva2) [inline src/Utility/Functions.C]: just checking. this was used to suppress output during ic parsing, you have reorganized so the output doesn't mess up ?

**Comment by timovanopstal** [comment]
Can I ask a naive question? Why? I'm sure there is solid design reasoning behind this further segmentation in the SIM hierarchy that I'm unaware of, hence the question. From where I stand, I already find it hard to understand the whole hierarchy, where to find which members and what not... But again, I'm sure there is a lot of gain too somewhere.

**Comment by akva2** [comment]
the idea is that this improves that situation. everything FEM model related is now in SIMmodel.

**Comment by kmokstad** [comment]
In the beginning the SIM hierachy was just (SIMinput -->) SIMbase --> SIM[123]D --> App-specific SIM's. As more and more functionality was added, SIMbase grew in size and became more and more difficult to keep tidy. So I introduced SIMoutput between SIMbase and SIM#D with all the output functionalities. But now, with the new model generation stuff, I find the time ready to put all the input/model generation stuff also in one place, therefore the new class. SIMbase now focuses on the computational tasks (plus utilities to access the model), which was its original purpose. I thought the new segregation to be an improvement and to help find things, not the opposite.

**Comment by timovanopstal** [comment]
Thanks for the explanations! Basically, a developer in IFEM needs to know which classes are in the SIM hierarchy and what these do, then it is easy to track data and functionality.

File: Apps/Common/SIMNodalConstraint.h (hunk 1 above)
Comment (kmokstad) [inline Apps/Common/SIMNodalConstraint.h]: It meant so, as I thought this was an obsolete option (not many apps use it, maybe only Elasticity and Poisson). It went out "in the dragsug" with this commit as I at an earlier stage had problem with the CharVec usage in this file. But now it seems no longer the case, likely due to other changes later, so I will revert this file change entirely (not related to the scope of this PR).

File: src/SIM/ModelGenerator.C (hunk 2 above)
Comment (kmokstad) [inline src/SIM/ModelGenerator.C]: sure, fixed in ec496f045b9f8a8d7a64c901a6c86d48c778a3a8. Think I got all of them.

File: src/SIM/Test/TestInitialConditions.C (hunk 11 above)
Comment (kmokstad) [inline src/SIM/Test/TestInitialConditions.C]: jo, men da måtte jeg endre alle ASSERT_EQ'ene også "." istedet for "->", så jeg tenke det var greit å minimere footprint av denne egentlige kosmetiske endringen.

File: src/Utility/Functions.C (hunk 12 above)
Comment (kmokstad) [inline src/Utility/Functions.C]: yes, the output is fixed in the reorganization (using IFEM::cout), so I assume it is OK in parallel also, although not checked.

Comment (akva2) [inline src/Utility/Functions.C]: wasn't parallel that caused the problem. however, i just checked, and printout is indeed fine now.

File: src/SIM/SIMmodel.h [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+class SIMmodel : public SIMbase
+{
+public:
+  //! \brief Struct holding information about an initial condition.
+  struct ICInfo
+  {
+    int file_level; //!< The time level for the field in the file
+    int geo_level;  //!< The time level for the geometry in the file
+    char basis;     //!< The basis to inject field into (for mixed)
```
Comment (akva2) [inline src/SIM/SIMmodel.h]: either default this to 1, or keep https://github.com/OPM/IFEM/blob/master/src/SIM/SIMbase.C#L476  breaks some tests otherwise.

**Comment by akva2** [comment]
Jeg ser ikke en eneste fordel med den løsningen? Pes og enkelt å gjøre feil,  for ingen gevinst noensted. hva har jeg gått glipp av?

**Comment by timovanopstal** [comment]
IFEM møte: suggested rename:
SIMinput --> SIMadmin
SIMmodel --> SIMinput

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-openfrac=15 ifem-advectiondiffusion=5 ifem-poroelasticity=10 ifem-thermoelasticity=6 ifem-stokes=42 ifem-navierstokes=41 ifem-fsi=7 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-openfrac=15 ifem-advectiondiffusion=5 ifem-poroelasticity=10 ifem-thermoelasticity=6 ifem-stokes=42 ifem-navierstokes=41 ifem-fsi=7 ifem-finitedeformation=3 please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-openfrac=15 ifem-advectiondiffusion=5 ifem-poroelasticity=10 ifem-thermoelasticity=6 ifem-stokes=42 ifem-navierstokes=41 ifem-fsi=7 ifem-finitedeformation=3 please

**Pull request merged**